### PR TITLE
Add `ResNetBackbone` and `ResNetImageClassifier`

### LINKED ISSUE #1750: add ResNetV1 and ResnetV2
This task is part of the KerasHub project ( Consolidating all of KerasCV and KerasNLP models in to one place)
Models from KerasCV will be added to KerasNLP in KerasNLP infrstructure style. The development is now under the keras-hub branch in KerasNLP - https://github.com/keras-team/keras-nlp/tree/keras-hub

Steps to complete this Task

- Source code : https://github.com/keras-team/keras-cv/tree/master/keras_cv/src/models/backbones/resnet_v1, https://github.com/keras-team/keras-cv/tree/master/keras_cv/src/models/backbones/resnet_v2
- Destination : A model folder here https://github.com/keras-team/keras-nlp/tree/keras-hub/keras_nlp/src/models
- Example PR : https://github.com/keras-team/keras-nlp/pull/1737, please follow the style from here and take a look at the comments on the PR for any design clarifications
- copy the model over
- If the model has multiple versions with small changes in architecture, they should be added as one model and the configuration should be handled in the args
- use KerasNLP test routines to update the tests
- for reformatting the code run shell/api_gen.sh, shell/format.sh and shell/lint.sh
- The backbone will be added as a model, any downstream task model will be under the same model folder.

Add either @divyashreepathihalli, @mattdangerw or @SamanehSaadat for review of any PR.

### PR DESCRIPTION
**Pull request by james77777778**
Closes #1750 

This PR largely follows #1737 but introduces some improvements:
- Supports both `"channels_last"` and `"channels_first"`
- Add support for mixed precision
- Improve the docstrings

The `preact` parameter in `ResNetBackbone` determines whether to instantiate ResNet or ResNetV2 (similar to `keras.applications.ResNet*`)
https://github.com/keras-team/keras/blob/b45be337c6156d90f220beba9f68eeb2e52e2b0d/keras/src/applications/resnet.py#L50

I have a small question regarding model implementation:
When creating a `Dense` layer in `ImageClassifier`, the `dtype_policy` is not configurable. I used `backbone.dtype_policy` as a workaround but it might be fragile.

The implementation has been verified by using the model weights from KerasCV.
Please refer to this colab: https://colab.research.google.com/drive/1HGktF-4TkTdTjlMVv7qr3D_ysspJXDmo?usp=sharing

@mattdangerw @divyashreepathihalli @SamanehSaadat

### CODE REVIEW

File: keras_nlp/src/models/resnet/resnet_backbone.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+    return x
+
+
+def apply_block(
+    x,
+    filters,
+    kernel_size=3,
+    stride=1,
+    conv_shortcut=False,
+    preact=False,
```
Comment (divyashreepathihalli) [inline keras_nlp/src/models/resnet/resnet_backbone.py:343]: same here with the arg name

File: keras_nlp/src/models/resnet/resnet_backbone_test.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        ("v1_basic_channels_last", False, "basic_block", "channels_last"),
+        ("v1_basic_channels_first", False, "basic_block", "channels_first"),
+        ("v1_block_channels_last", False, "block", "channels_last"),
+        ("v2_basic_channels_last", True, "basic_block", "channels_last"),
+        ("v2_basic_channels_first", True, "basic_block", "channels_first"),
+        ("v2_block_channels_last", True, "block", "channels_last"),
+    )
+    def test_backbone_basics(self, preact, block_type, data_format):
+        if (
+            backend.backend() == "tensorflow"
```
Comment (divyashreepathihalli) [inline keras_nlp/src/models/resnet/resnet_backbone_test.py:46]: lets not set the backend. This way the tests will run on all backends

**Review by divyashreepathihalli — APPROVED** [review]
Thanks for the PR! Looks great! I have 2 minor NIT comments. Otherwise LGTM, subject to  Matt's approval.

File: keras_nlp/src/models/resnet/resnet_backbone.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+import keras
+from keras import layers
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.backbone import Backbone
+from keras_nlp.src.utils.keras_utils import standardize_data_format
+
+
+@keras_nlp_export("keras_nlp.models.ResNetBackbone")
+class ResNetBackbone(Backbone):
```
Comment (mattdangerw) [inline keras_nlp/src/models/resnet/resnet_backbone.py:23]: Do we think this makes more sense as one architecture or two? We can do both here. Either `ResNetBackbone` and `ResNet2Backbone` or the one like we have here.

Most of the args makes sense here, but `block_type: "block"` vs `block_type: "basic_block"` is a little hard to understand. But also night to have less duplication.

Basically it's on us to define what constitutes a new architecture.

Comment (divyashreepathihalli) [reply keras_nlp/src/models/resnet/resnet_backbone.py:23]: how about providing different architectures as main backbone subclasses with built in configs?

Comment (mattdangerw) [reply keras_nlp/src/models/resnet/resnet_backbone.py:23]: Talked offline. Let's just stick with the one class for now, nice and simple.

If we can think of something still short that's a little more descriptive than `block_type: "block"` and `block_type: "basic_block"` let's go for it, but I think this is fine as is.

**Review by mattdangerw — APPROVED** [review]
Left some minor comments, but all good on my end once addressed. @divyashreepathihalli I'll leave the final merge to you!

File: keras_nlp/src/models/resnet/resnet_backbone_test.py (hunk 2 above)
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_backbone_test.py:46]: but tensorflow cpu can't perform convolution with the `channels_first` format
I can change it to detect whether the gpu is available or not.

File: keras_nlp/src/models/resnet/resnet_backbone_test.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    @parameterized.named_parameters(
+        ("v1_basic_channels_last", False, "basic_block", "channels_last"),
+        ("v1_basic_channels_first", False, "basic_block", "channels_first"),
+        ("v1_block_channels_last", False, "block", "channels_last"),
+        ("v2_basic_channels_last", True, "basic_block", "channels_last"),
+        ("v2_basic_channels_first", True, "basic_block", "channels_first"),
+        ("v2_block_channels_last", True, "block", "channels_last"),
+    )
+    def test_backbone_basics(self, preact, block_type, data_format):
+        if (
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_backbone_test.py]: Added `run_vision_backbone_test` with the `run_data_format_check` option.
This should simplify the code for future vision models.

File: keras_nlp/src/models/resnet/resnet_backbone.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        stackwise_num_filters,
+        stackwise_num_blocks,
+        stackwise_num_strides,
+        block_type,
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_backbone.py:101]: Updated.
The new signature:
```python
    def __init__(
        self,
        stackwise_num_filters,
        stackwise_num_blocks,
        stackwise_num_strides,
        block_type,
        use_pre_activation=False,
        include_rescaling=True,
        input_image_shape=(None, None, 3),
        pooling="avg",
        data_format=None,
        dtype=None,
        **kwargs,
    ):
```

I think we can change `input_image_shape` from a fixed `(224, 224, 3)` to `(None, None, 3)`. This should be more user-friendly.

File: keras_nlp/src/models/resnet/resnet_backbone.py (hunk 3 above)
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_backbone.py:23]: Updated to `block_type: "bottleneck_block"` and `block_type: "basic_block"`.
This should be consistent with the original paper: https://arxiv.org/abs/1512.03385

File: keras_nlp/src/models/resnet/resnet_backbone_test.py (hunk 2 above)
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_backbone_test.py]: Added a check for TF CPU in newly introduced `run_vision_backbone_test`

**Comment by james77777778** [comment]
Another question:
Should we support pyramid feature outputs in vision backbones?
This is crucial for any downstream algorithms (object detection, segmentation, ...)

https://github.com/keras-team/keras-cv/blob/5cfe7c3d8f2d758b91bbfcedf7f0a29931236c68/keras_cv/src/models/backbones/backbone.py#L153-L176

**Comment by divyashreepathihalli** [comment]
> Another question: Should we support pyramid feature outputs in vision backbones? This is crucial for any downstream algorithms (object detection, segmentation, ...)
> 
> https://github.com/keras-team/keras-cv/blob/5cfe7c3d8f2d758b91bbfcedf7f0a29931236c68/keras_cv/src/models/backbones/backbone.py#L153-L176

This sounds like a good idea!

**Review by divyashreepathihalli — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**